atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/sweep/GetCellTimestamps.java

```diff
@@ -63,14 +76,51 @@ public GetCellTimestamps(
      * row.
      */
     private void fetchBatchOfTimestamps() {
-        fetchAllTimestampsBeginningAtStartRow();
+        fetchBatchOfTimestampsBeginningAtStartRow();
+
         fetchRemainingTimestampsForLastRow();
     }
 
-    private void fetchAllTimestampsBeginningAtStartRow() {
-        List<CellWithTimestamp> batch = cqlExecutor.getTimestamps(tableRef, startRowInclusive, batchHint);
+    /**
+     * The complexity in this method is due to a desire to avoid scanning over large numbers of tombstoned rows.
+     * <p>
+     * Details: An unbounded CQL query will scan over an unbounded number of tombstoned rows in an attempt to find
+     * either the end of the table, or the desired number of results. This can lead to timeouts if there are large
+     * segments of tombstoned rows (which can happen with THOROUGH sweep). Thrift, on the other hand, will actually
+     * return the tombstoned rows (they will just be empty KeySlices), which allows us to page over them. So, we first
+     * execute a thrift range scan to determine an upper bound on the CQL range scan, and page over the data until we
+     * find some live cells.
+     * <p>
+     * This is a hack (and a perf hit) until we have a better solution to avoid scanning massive numbers of
+     * tombstones (ideally this will involve catching some exception and reducing the range appropriately).
+     */
+    private void fetchBatchOfTimestampsBeginningAtStartRow() {
+        Optional<byte[]> rangeEnd;
+        for (byte[] rangeStart = startRowInclusive; timestamps.isEmpty(); rangeStart = rangeEnd.get()) {
+            rangeEnd = determineSafeRangeEnd(rangeStart);
+            if (!rangeEnd.isPresent()) {
+                return;
+            }
+
+            List<CellWithTimestamp> batch = cqlExecutor.getTimestamps(tableRef, rangeStart, rangeEnd.get(), batchHint);
+            timestamps.addAll(batch);
+        }
+    }
 
-        timestamps.addAll(batch);
+    private Optional<byte[]> determineSafeRangeEnd(byte[] rangeStart) {
+        KeyRange keyRange = new KeyRange().setStart_key(rangeStart).setEnd_key(new byte[0]).setCount(batchHint);
+        SlicePredicate slicePredicate = SlicePredicates.create(SlicePredicates.Range.ALL, SlicePredicates.Limit.ONE);
```

Comment:
Does the `Limit.ONE` effectively mean that the `List<KeySlice>` will just contain up to `batchHint` row keys with 0-1 columns?

Reply:
Do we need to be concerned about loading values with this query?

Reply:
Yes, we need to be concerned with loading values here. I wonder if it's possible to execute a range slice query with a limit of zero..